llama_index/indices/knowledge_graph/retriever.py

```diff
@@ -147,6 +147,10 @@ def _retrieve(
                         # set. This helps introduce more global knowledge into the
                         # query. While it's more expensive, thus to be turned off
                         # by default, it can be useful for some applications.
+
+                        # TODO: we should a keyword-node_id map in IndexStruct, so that
```

Comment:
When I am thinking of @jerryjliu 's idea and going through the existing retriever, I realized this is too costly when global explore is enabled and thus put this TODO here, @logan-markewich what do you think, please?

Reply:
Sorry, I meant we already have KG.table(keyword->node_id), so just introduce a KG.<reverse_table>(node_id->keyword). in data_structs.py only adds a dict setting during `KG.add_node()`?

```python
    def add_node(self, keywords: List[str], node: Node) -> None:
        """Add text to table."""
        node_id = node.get_doc_id()
        for keyword in keywords:
            if keyword not in self.table:
                self.table[keyword] = set()
            self.table[keyword].add(node_id)
```

Reply:
@wey-gu i see, this makes sense. could you point me to the place where this doubles calls to the LLM for keyword extraction?

Reply:
> @wey-gu i see, this makes sense. could you point me to the place where this doubles calls to the LLM for keyword extraction?

This line calls LLM to extract keywords from nodes( `self._get_keywords()`, in line 154 after this comment. By default, it's not called as `self.use_global_node_triplets` is off by default. :)

I can optimize this in the next PR.

```python
                        extended_subjs = self._get_keywords(
                            self._docstore.get_node(node_id).get_text()
                        )
```